Dystopia excnahge integration

bogdoslavik opened the pull request:
Hello!

This is draft integration with one question.
I trying to integrate Dystopia Exchange https://www.dystopia.exchange/
It is UniswapV2 fork with two types of pools:
1st - uniswap like, and
2nd - stable pools with x3y+y3x curve

I decided to split one integration to two virtual dexes:

1. `dystopia.ts` - uniswap2 like pools, and it implemented with no issues.
2. `dystopia-stable.ts` - stable pools

Stable pools have same interfaces as uniswap2 pools, BUT have another formula to calculate amount out (pairs have external [`function getAmountOut`](https://github.com/dystopia-exchange/dystopia-contracts/blob/0bf82cef3d94f1e35c2dd0dc84a8db246cef3ca4/contracts/base/core/DystPair.sol#L479) )

For `simpleSwap `I've extended Dystopia router with swap function, with same interface like in `NewUniswapV2ExchangeRouter`, and `simpleSwap `sucessfully forwarded trough this router.

But `multiSwap` and `megaSwap` uses `swapOnUniswapV2Fork` in `PolygonAdapter01`, that uses `NewUniswapV2Lib.getAmountOut` to calculate amount out.

To complete the integration, we need same function, like `swapOnUniswapV2Fork`, but that will use `DystPair.getAmountOut` instead of `NewUniswapV2Lib.getAmountOut`.

How we can add such function to your swap adapter?
I can prepare solidity code of the new function, but I need assistance from you to integrate it to your contract.

Thank you!

bogdoslavik commented:
Hello! I have merged with upstream.
Dystopia tests passed
DystopiaStable - only simple swap passed, multiswap and mega swap needs adapter contract extension from your side (see details above).

Thank you!

bogdoslavik commented on src/dex/index.ts:
```diff
@@ -40,7 +40,8 @@ import { MakerPsm } from './maker-psm/maker-psm';
 import { Platypus } from './platypus/platypus';
 import { GMX } from './gmx/gmx';
-import { WooFi } from './woo-fi/woo-fi';
+import { Dystopia } from './uniswap-v2/dystopia';
+import { DystopiaStable } from './uniswap-v2/dystopia-stable';
```
Sorry, looks like I have mistaked when merging upstream

bogdoslavik replied:
Sorry, looks like I do mistage when merging...

bogdoslavik commented on src/dex/uniswap-v2/uniswap-v2.ts:
```diff
@@ -85,7 +85,7 @@ export type UniswapV2Pair = {
 };
 
-const subgraphTimeout = 10 * 1000;
+export const subgraphTimeout = 10 * 1000;
```
I've moved it to the` /src/constants.ts` as same constant aso used at the balancer-v2 and `kyberdmm.ts`

bogdoslavik commented on src/dex/uniswap-v2/dystopia-stable-integration.test.ts:
```diff
@@ -0,0 +1,64 @@
+import { DummyDexHelper } from '../../dex-helper';
+import { Network, SwapSide } from '../../constants';
+import { DystopiaStable } from './dystopia-stable';
+// @ts-ignore
```
I have this warning in my IDE (WebStorm) for all files imported from: 

TS6059: File 'C:/Users/Dev/Projects/paraswap-dex-lib-dystopia/tests/utils.ts' is not under 'rootDir' 'C:/Users/Dev/Projects/paraswap-dex-lib-dystopia/src'. 'rootDir' is expected to contain all source files.

Possible tsconfig.json needs some additional setting, but I did not found what.

![image](https://user-images.githubusercontent.com/28448359/172305837-1db8422e-00df-4dd8-9f49-53d8a1860697.png)

bogdoslavik replied:
I have removed `// @ts-ignore`

bogdoslavik replied:
Found the solution:
I have extended tsconfig with

{
  "compilerOptions": {
    "rootDirs": [
      "src",
      "tests"
    ]
  }
}

bogdoslavik commented on src/dex/uniswap-v2/uniswap-v2.ts:
```diff
@@ -45,11 +45,11 @@ import { Contract } from 'web3-eth-contract';
 import { UniswapV2Config, Adapters } from './config';
 import { BI_MAX_UINT } from '../../bigint-constants';
 
-const RESERVE_LIMIT = 2n ** 112n - 1n;
+export const RESERVE_LIMIT = 2n ** 112n - 1n;
```
Importing order do not give me move `RESERVE_LIMIT` to `src/dex/uniswap-v2/constants.ts`: when Ido that, then I get error in `src/dex/uniswap-v2/mdex.ts:22:33`: 

> TypeError: Class extends value undefined is not a constructor or null

Because `constants.ts` imports few uniswap fork dex configs what extends UniswapV2 class that is not yet compiled..

Verisana replied:
I see. Thanks for pointing it out. Then let's leave it as it is

bogdoslavik commented on src/dex/uniswap-v2/dystopia.ts:
```diff
@@ -0,0 +1,224 @@
+  subgraphURL:
+    'https://api.thegraph.com/subgraphs/name/dystopia-exchange/dystopia-v2',
+  factoryAddress: '0x1d21Db6cde1b18c7E47B0F7F42f4b3F68b9beeC9',
+  // ParaSwap-compatible Router with stable pools support
+  router: '0x0E98A8e5ca6067B98d10Eb6476ec30E232346402',
```
Not sure what it is adds gas overhead, as it is just integrated in to Dystopia router:  https://polygonscan.com/address/0x0e98a8e5ca6067b98d10eb6476ec30e232346402#code#F1#L638

bogdoslavik commented on src/dex/uniswap-v2/dystopia-stable.ts:
```diff
@@ -0,0 +1,248 @@
+    return returnAmount;
+  }
+
+  // TODO Do we have to implement getBuyPrice? (buy side do not used in tests)
```
Can we omit buy side for now and integrate sell only? 

As I have no ready formula for stable pools swaps and it can take some time to properly made it, and we want to integrate  ASAP. 

I guess sell side is used in most cases..

Verisana replied:
Ok, I discussed this request with @erez-paraswap and received his approval. We can integrate first with `SELL` and `BUY` later.

As for how is it better to do, please check, for example `Nerve`:
https://github.com/paraswap/paraswap-dex-lib/blob/f5d7319fc6cc3cba7c888b2b18027dd41e891328/src/dex/nerve/nerve.ts#L172

In general: for `getPricesVolume` and `getPoolIfentifiers` you check the side and return `[]`, as for building params, you should throw an `Error` in `getSimpleParam` and `getAdapterParam`.

bogdoslavik commented:
Thank you for your comments! 
I have resolved almost all. Hope myPR now little better )

My collegues want to integrate asap, so please lt's not implement the buy side for now.

Verisana commented:
Looks good 👍 

Thank you for the changes. Regarding my fourth comment. I discussed with the team and it is better to put one exchange into one `Dystopia` instance. Not separating `Dystopia` and `DystopiaStable`.

You can check that `BalancerV2` has different pools too, but there is only one `BalancerV2` exchange. We keep: one exchange protocol include different pools.

So, for `Dystopia` it will be two different pools inside it. You can differentiate between types from config. And when requesting `getPoolIdentifiers` you should search in all pools (all types) and return the ones that contain `srcToken` and `destToken`. I didn't check it, but I suppose that all pools has unique pairs without dublicate pools. In that case the request should return only one pool or nothing.

I kindly remind about the onchain price checking. It is very important.

After we complete this part, we can go to adapter implementation. We will do it on our side.

And when it is all done, my collegue will have some final checks and if it all ok, we will be glad to merge `Dystopia` exchange :)

bogdoslavik commented:
@Verisana hello again! )

I've joined Dystopia and Dystopia Stable into one Dystopia and moved uniswap-like and stable pools calcutaions to the separate files.

For now all tests passed, but with fluctuations (sometimes it can not pass, after re-run it sucessfully passed). Not sure why yet. :(
Will try dive deeper tomorrow morning.

> I kindly remind about the onchain price checking. It is very important.

What I missed out and not implemented? Could you please point me?

Verisana commented:
@bogdoslavik 
Thanks for the changes 👍 
I will check after all issues are resolved.

> For now all tests passed, but with fluctuations (sometimes it can not pass, after re-run it sucessfully passed). Not sure why yet. :(
If it e2e tests, it may be due to the reasons unrelated to implementation. For example, `header not found` or `timeout` issues. If you give more context on the error, we can evaluate, if it related or not. Unfortunately public RPCs not very reliable.

> What I missed out and not implemented? Could you please point me?
Can you check please this message in Telegram:
https://t.me/c/1631443647/51

Maybe you missed it because it was in the time when you were travelling. If it didn't answer your question fully, come back and we can discuss what part needs more detalization

bogdoslavik commented:
@Verisana hi!

I've enlarged FETCH_POOL_PRICES_TIMEOUT and all tests now passed.
Also I have added checkOnChainPricing to the integration tests - yep, looks like I missed it while was travelling.  

Guess PR ready for review )

bogdoslavik commented:
I found, what function [getPrices at local-paraswap-sdk](https://github.com/dystopia-exchange/paraswap-dex-lib/blob/e121025b62ad05751d80e9ee144c0a10c77f0756/src/implementations/local-paraswap-sdk.ts#L95)  

uses only first pool, but we may have 2 pools for one pair,
so I have added sorting, so for now pool with larger unit price will be first.

And now stable pools tests failed, as expected 
(except SimpleSwap as I have added own router to configuration that supports stable pools).

bogdoslavik commented:
@Verisana please see my previous comment

Verisana commented:
@bogdoslavik 
Thanks for all changes. Looks good to me 👍 

Today I will handle adapter and test one more time. If it all ok, we can put it to prod :)
Created my preparation [PR#121](https://github.com/paraswap/paraswap-dex-lib/pull/121) and merged all your changes there
